nvram_export.1
.TH nvram_export 1 "Sep 07, 2006"
.SH NAME
nvram_export \- Cisco 7200 NVRAM configuration export
.SH SYNOPSIS
.B nvram_export
.I nvram_file
.I config_file
.br
.SH DESCRIPTION
export configuration from dynamips nvram file.
.SH REPORTING BUGS
.br
Please send bug reports to 
.UR https://github.com/GNS3/dynamips-community/issues
.UE
.SH SEE ALSO
.br
\fBdynamips\fP(1), \fBhypervisor_mode\fP(7)
.br
.UR http://www.gns3.net/dynamips/
.UE
.br
.UR http://forum.gns3.net/
.UE
.br
.UR https://github.com/GNS3/dynamips-community
.UE
.br
.SH OLD WEBSITES
.UR http://www.ipflow.utc.fr/index.php/
.UE
.br
.UR http://www.ipflow.utc.fr/blog/
.UE
.br
.UR http://hacki.at/7200emu/index.php
.UE
.SH AUTHOR
\fBdynamips\fP was written by Fabien Devaux, Christophe Fillot (cf@utc.fr)
and MtvE. This manual page was written by Erik Wenzel <erik@debian.org>
for the Debian GNU/Linux system (but may be used by others). This manual 
page was updated for the dynamips-community fork by Flávio J. Saraiva 
<flaviojs2005@gmail.com>.



ChangeLog
/*
 * Cisco 7200/3600/2691 simulation platform.
 */

15-Aug-2005, the project is starting.

18-Aug-2005
-----------
  - Added support for: sll, addu/subu.
  - Fixed bugs with b... branch instructions (bad computation of new pc).
  - Added patch recording for blocks.
  - A sequence of MIPS instructions (without memory access instructions)
    from IOS was successfully run ! (need to check correctness).

22-Aug-2005
-----------
  - Implemented a basic memory subsystem with an mmaped file.
  - Added memory access instructions (lb,lbu,lh,lhu,lw,lwu,ld,sb,sh,sw,sd).
  - Added a cross-compiled test program.

23-Aug-2005
-----------
  - Fixed slti, sltiu, slt instructions.
  - Fixed memory endianness.
  - Fixed host memory addressing (bad haddr computation).

24-Aug-2005
-----------
  - Added a generic memory address lookup function.
  - Added a minimal ELF loader.

28-Aug-2005
-----------
  - ELF loader now loads files page per page.
  - Added device support (with RAM support).
  - Added MTS32 support for mapping of 4 Kb pages.
  - Added MTS32 support for device mapping in kernel mode and generic
    virtual to physical mapping (preparation for TLB).
  - Added a dummy console driver.
  - Added mult,multu,div,divu instructions.

29-Aug-2005
-----------
  - Added a Red-Black trees module.
  - Block locator uses RB tree module.
  - It is now possible to run a program with multiple functions!
  - ELF loader now support multiple program headers and entry point is
    now used.
  - Fixed x86 buffer adjustment.
  - Modified console to be at the same address than c7200 DUART (to test
    microcode).
  - Memory access functions now update PC.
  - Added sra, srl, sllv, srlv, srav instructions.

30-Aug-2005
-----------
  - Added dsrl32, dsll32, dsra, dsrav, dsra32 instructions.
  - Added bgtz, blez, bgez instructions.
  - Fixed sltu instruction.
  - ELF Entry point is now sign-extended.
  - Added blezl, bgtzl instructions.
  - Fixed the MIPS scanner to use unsigned 64-bits integers for addresses.
  - Added unknown opcode management (basic, it just print the unhandled
    opcode value and continues).
  - Added cache instruction.
  - Added bltzal, bltzall, bgezal, bgezall instructions.

31-Aug-2005
-----------
  - Added System coprocessor (CP0) definitions.
  - Added tlbr instruction.
  - Added basic tlbwi instruction (doesn't change the MTS mappings now).
  - Added dmfc0, dmtc0, mfc0, mtc0 instructions.
  - Added TLB dump functions.
  - Added dummy pref/prefi instructions.
  - Added basic exception trigger.
  - Added syscall, break instructions.

01-Sep-2005
-----------
  - Fixed exception trigger.
  - Added dmfc1, dmtc1, mfc1, mtc1, sdc1 instructions.
  - Fixed the break instruction.
  - Modified the memory handling to be more friendly with instructions
    like sdc1 and for more safety in case of badly written device drivers.
  - Added ldc1 instruction.

02-Sep-2005
-----------
  - Fixed sltiu/sltu instructions that were doing signed comparisons
    instead of unsigned comparisons (mis-use of jit).
  - Added lwr/lwl, swr/swl, ldl/ldr, sdr/sdl, ll/sc instructions
    (not tested).
  - Added PCI bus management.
  - Added GT64010 (PCI controller) device.
  - Added DEC 21050 (PCI bridge) device.
  - Added NVRAM.

03-Sep-2005
-----------
  - Beginning of work on IRQs.

04-Sep-2005
-----------
  - Timer IRQ should be OK.
  - Added C7200 bay management and Midplane FPGA.
  - Fixed the JR instruction which was broken in a case like that:
	lwu k1,0(k1)
	jr k1
	move k1,zero

05-Sep-2005
-----------
  - Fixed jalr (as jr) instruction.
  - Added RAM "aliasing".
  - Added SRAM management.
  - Fixed NVRAM size.
  - Fixed ldr, swl/swr instructions.
  - Added c7200 IOcard.
  - Count register is now properly handled.
  - Added basic functions to set/clear interrupts.
  - Modified the console driver to trigger DUART interrupt.
  - IOS now starts!

06-Sep-2005
-----------
  - Added "move" virtual instruction. It correspond to "addu" with rt=0.
    Here it avoids an useless add.
  - Added b/bal virtual instructions.
  - Packed NVRAM.

08-Sep-2005
-----------
  - Added Dallas DS1620 temperature sensors.
  - Added voltage sensors.
  - Power supplies are now faked.
  - IOS Environmental Monitor doesn't try to shutdown the router anymore!
  - Device list is now displayed at startup.

18-Sep-2005
-----------
  - Added Address Error Load/Save, TLB Load/Save exceptions.
    TODO: Branch Delay Slot management.
  - Modified the default value in config register in CP0 to set appropriate
    cache sizes. Now it is possible to boot compressed images directly!
  - Added the teq/teqi instructions.
  - Hmm, seen some IOS images (k9?) that seem to use the FPU (CP1). This
    does not prevent the image to boot, but some features will probably
    be not functional.

19-Sep-2005
-----------
  - Tweaked the console driver to generate a dummy interrupt. It avoids
    the console problems.
  - Fixed the serial driver mueslix PCI vendor/device code.
  - The midplane FPGA now returns voltages based on virtual PA EEPROM
    presence. Up to 6 PA-8T were successfully emulated!

20-Sep-2005
-----------
  - Included amd64-codegen.h with a (near empty) translator.
  - Some code factorization.

21-Sep-2005
-----------
  - Added CPU state save/restore on disk.
  - Sending SIGQUIT to the process dumps the MIPS64 VM state on disk.
    (only CPU state is saved for now).
    Device state is not saved, so this does not work with booted IOS images
    for now (it works for an IOS image being self-decompressing, though).
  - Some code cleanups.

23-Sep-2005
-----------
  - Fixed "Dirty" bit in TLB entries.
  - Extended the number of TLB entries from 48 to 64.
  - Added tlbp instruction.
  - Modified the nmc93c46 eeprom driver to allow more than 32 bytes reading.

24-Sep-2005
-----------
  - Added a symbol table loader.

26-Sep-2005
-----------
  - Added support for writing in PCI registers
  - Added a RAW file loader.

common/nvram_export.c
/*
 * Cisco C7200 Simulation Platform.
 * Copyright (c) 2005,2006 Christophe Fillot.  All rights reserved.
 *
 * Extract IOS configuration from a NVRAM file (standalone tool)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <errno.h>

#include "cpu.h"
#include "dynamips.h"
#include "memory.h"
#include "device.h"
#include "dev_c7200.h"
#include "dev_c3600.h"
#include "dev_c2600.h"
#include "dev_c1700.h"

#define PLATFORM_C7200	0x37323030
#define PLATFORM_C3600	0x33363030
#define PLATFORM_C2600	0x43323630
#define PLATFORM_C1700	0x43313730

/* Export configuration from 7200 NVRAM */
int dev_nvram_7200_export_config(FILE *nvram_fd,FILE *cfg_fd)
{
   m_uint32_t tag,start,end,len,clen,nvlen;
   char buffer[512];
   int res = -1;

   fseek(nvram_fd,C7200_NVRAM_ROM_RES_SIZE+6,SEEK_SET);
   fread(&tag,sizeof(tag),1,nvram_fd);
   if (ntohl(tag) != 0xF0A5ABCD) {
      fprintf(stderr,"NVRAM: Unable to find IOS tag (tag=0x%8.8x)!\n",
              ntohl(tag));
      goto done;
   }

   fseek(nvram_fd,0x06,SEEK_CUR);
   fread(&start,sizeof(start),1,nvram_fd);
   fread(&end,sizeof(end),1,nvram_fd);
   fread(&nvlen,sizeof(nvlen),1,nvram_fd);
   start = htonl(start) + 1;
   end   = htonl(end);
   nvlen = htonl(nvlen);

   if ((start <= C7200_NVRAM_ADDR) || (end <= C7200_NVRAM_ADDR) || 
       (end <= start)) 
   {
      fprintf(stderr,"NVRAM: invalid configuration markers "
              "(start=0x%x,end=0x%x).\n",start,end);
      goto done;
   }
   
   clen = len = end - start;
   if ((clen + 1) != nvlen) {
      fprintf(stderr,"NVRAM: invalid configuration size (0x%x)\n",nvlen);
      goto done;
   }

   start -= C7200_NVRAM_ADDR;
   fseek(nvram_fd,start,SEEK_SET);

   while(len > 0) {
      if (len > sizeof(buffer))
         clen = sizeof(buffer);
      else
         clen = len;

      fread(buffer,clen,1,nvram_fd);
      fwrite(buffer,clen,1,cfg_fd);
      len -= clen;
   }

   res = 0;
 done:
   return res;
}

/* Export configuration from 3600 NVRAM */
int dev_nvram_3600_export_config(FILE *nvram_fd,FILE *cfg_fd)
{
   m_uint32_t tag,start,end,len,clen,nvlen;
   char buffer[512];
   int res = -1;

   fseek(nvram_fd,C3600_NVRAM_ROM_RES_SIZE+6,SEEK_SET);
   fread(&tag,sizeof(tag),1,nvram_fd);
   if (ntohl(tag) != 0xF0A5ABCD) {
      fprintf(stderr,"NVRAM: Unable to find IOS tag (tag=0x%8.8x)!\n",
              ntohl(tag));
      goto done;
   }

   fseek(nvram_fd,0x06,SEEK_CUR);
   fread(&start,sizeof(start),1,nvram_fd);
   fread(&end,sizeof(end),1,nvram_fd);
   fread(&nvlen,sizeof(nvlen),1,nvram_fd);
   start = htonl(start) + 1 + C3600_NVRAM_ADDR + C3600_NVRAM_ROM_RES_SIZE + 8;
   end   = htonl(end) + C3600_NVRAM_ADDR + C3600_NVRAM_ROM_RES_SIZE + 8;
   nvlen = htonl(nvlen);

   if ((start <= C3600_NVRAM_ADDR) || (end <= C3600_NVRAM_ADDR) || 
       (end <= start)) 
   {
      fprintf(stderr,"NVRAM: invalid configuration markers "
              "(start=0x%x,end=0x%x).\n",start,end);
      goto done;
   }
   
   clen = len = end - start;
   if ((clen + 1) != nvlen) {
      fprintf(stderr,"NVRAM: invalid configuration size (0x%x)\n",nvlen);
      goto done;
   }

   start -= C3600_NVRAM_ADDR;
   fseek(nvram_fd,start,SEEK_SET);

   while(len > 0) {
      if (len > sizeof(buffer))
         clen = sizeof(buffer);
      else
         clen = len;

      fread(buffer,clen,1,nvram_fd);
      fwrite(buffer,clen,1,cfg_fd);
      len -= clen;
   }

   res = 0;
 done:
   return res;
}

/* copied from dev_c2600.c */

static inline m_uint8_t c2600_nvram_read_byte(FILE *nvram_fd,u_int offset)
{
   m_uint32_t tag;

   if ( offset != 0 )
     fseek(nvram_fd, offset << 2, SEEK_SET);
   fread(&tag,sizeof(tag),1,nvram_fd);

   return (m_uint8_t)(tag&255);
}

static m_uint32_t c2600_nvram_read32(FILE *nvram_fd,u_int offset)
{
   m_uint32_t val;

   if ( offset != 0 )
     fseek(nvram_fd, offset << 2, SEEK_SET);

   val = c2600_nvram_read_byte(nvram_fd, 0) << 24 |
         c2600_nvram_read_byte(nvram_fd, 0) << 16 |
         c2600_nvram_read_byte(nvram_fd, 0) << 8 |
         c2600_nvram_read_byte(nvram_fd, 0);

   return(val);
}

/* Export configuration from 2600 NVRAM */
int dev_nvram_2600_export_config(FILE *nvram_fd,FILE *cfg_fd)
{
   m_uint32_t tag,start,len,clen,nvlen;
   char buffer[512];
   int res = -1;
   u_int i;
 
   tag = c2600_nvram_read32(nvram_fd, C2600_NVRAM_ROM_RES_SIZE+6);

   if (tag != 0xF0A5ABCD) {
      fprintf(stderr,"NVRAM: Unable to find IOS tag (tag=0x%8.8x)!\n",
              ntohl(tag));
      goto done;
   }

   start = c2600_nvram_read32(nvram_fd, C2600_NVRAM_ROM_RES_SIZE+0x10);
   nvlen = c2600_nvram_read32(nvram_fd, C2600_NVRAM_ROM_RES_SIZE+0x18);

   if ((start <= 0) || (start+nvlen <= 0) || 
       (nvlen <= 0)) 
   {
      fprintf(stderr,"NVRAM: invalid configuration markers "
              "(start=0x%x,nvlen=0x%x).\n",start,nvlen);
      goto done;
   }
   
   start += C2600_NVRAM_ROM_RES_SIZE + 0x08 + 1;
   
   clen = len = nvlen-1;

   fseek(nvram_fd,start<<2,SEEK_SET);

   while(len > 0) {
      if (len > sizeof(buffer))
         clen = sizeof(buffer);
      else
         clen = len;

	  for ( i = 0; i < clen; i++ )
	    buffer[i] = c2600_nvram_read_byte ( nvram_fd, 0 );
      fwrite(buffer,clen,1,cfg_fd);
      len -= clen;
   }

   res = 0;
 done:
   return res;
}

/* Export configuration from 1700 NVRAM */
int dev_nvram_1700_export_config(FILE *nvram_fd,FILE *cfg_fd)
{
   m_uint32_t tag,start,end,len,clen,nvlen;
   char buffer[512];
   int res = -1;

   fseek(nvram_fd,C1700_NVRAM_ROM_RES_SIZE+6,SEEK_SET);
   fread(&tag,sizeof(tag),1,nvram_fd);
   if (ntohl(tag) != 0xF0A5ABCD) {
      fprintf(stderr,"NVRAM: Unable to find IOS tag (tag=0x%8.8x)!\n",
              ntohl(tag));
      goto done;
   }

   fseek(nvram_fd,0x06,SEEK_CUR);
   fread(&start,sizeof(start),1,nvram_fd);
   fread(&end,sizeof(end),1,nvram_fd);
   fread(&nvlen,sizeof(nvlen),1,nvram_fd);
   start = htonl(start) + 1 + C1700_NVRAM_ADDR + C1700_NVRAM_ROM_RES_SIZE + 8;
   end   = htonl(end) + C1700_NVRAM_ADDR + C1700_NVRAM_ROM_RES_SIZE + 8;
   nvlen = htonl(nvlen);

   if ((start <= C1700_NVRAM_ADDR) || (end <= C1700_NVRAM_ADDR) || 
       (end <= start)) 
   {
      fprintf(stderr,"NVRAM: invalid configuration markers "
              "(start=0x%x,end=0x%x).\n",start,end);
      goto done;
   }
   
   clen = len = end - start;
   if ((clen + 1) != nvlen) {
      fprintf(stderr,"NVRAM: invalid configuration size (0x%x)\n",nvlen);
      goto done;
   }

   start -= C1700_NVRAM_ADDR;
   fseek(nvram_fd,start,SEEK_SET);

   while(len > 0) {
      if (len > sizeof(buffer))
         clen = sizeof(buffer);
      else
         clen = len;

      fread(buffer,clen,1,nvram_fd);
      fwrite(buffer,clen,1,cfg_fd);
      len -= clen;
   }

   res = 0;
 done:
   return res;
}

/* Export configuration from NVRAM */
int dev_nvram_export_config(char *nvram_filename,char *cfg_filename)
{
   m_uint32_t platform;
   FILE *nvram_fd,*cfg_fd;
   char platform_string[5];
   int res = -1;

   if (!(nvram_fd = fopen(nvram_filename,"r"))) {
      fprintf(stderr,"Unable to open NVRAM file '%s'!\n",nvram_filename);
      return(-1);
   }

   if (!(cfg_fd = fopen(cfg_filename,"w"))) {
      fprintf(stderr,"Unable to create config file '%s'!\n",cfg_filename);
      return(-1);
   }

   fseek(nvram_fd,0x200,SEEK_SET);
   fread(&platform,sizeof(platform),1,nvram_fd);

   switch(ntohl(platform)) {
      case PLATFORM_C7200:
	 res = dev_nvram_7200_export_config(nvram_fd,cfg_fd);
	 break;
      case PLATFORM_C3600:
	 res = dev_nvram_3600_export_config(nvram_fd,cfg_fd);
	 break;
      case PLATFORM_C2600:
	 res = dev_nvram_2600_export_config(nvram_fd,cfg_fd);
	 break;
      case PLATFORM_C1700:
	 res = dev_nvram_1700_export_config(nvram_fd,cfg_fd);
	 break;
      default:
	 memcpy(platform_string,&platform,4);
	 platform_string[4] = '\0';
	 fprintf(stderr,"NVRAM: unknown platform %s (platform=0x%x)!\n",
	         platform_string,ntohl(platform));
   }
 
   fclose(nvram_fd);
   fclose(cfg_fd);
   return(res);
}

int main(int argc,char *argv[])
{
   printf("Cisco 7200/3600 NVRAM configuration export.\n");
   printf("Copyright (c) 2006 Christophe Fillot.\n\n");

   if (argc != 3) {
      fprintf(stderr,"Usage: %s nvram_file config_file\n",argv[0]);
      exit(EXIT_FAILURE);
   }

   if (!dev_nvram_export_config(argv[1],argv[2]))
      printf("Configuration written to %s\n",argv[2]);

   return(0);
}
